format-common/src/main/java/io/cdap/plugin/format/SchemaDetector.java

```diff
@@ -0,0 +1,160 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format;
+
+import com.google.common.base.Strings;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.validation.FormatContext;
+import io.cdap.cdap.etl.api.validation.ValidatingInputFormat;
+import io.cdap.plugin.common.batch.JobUtils;
+import io.cdap.plugin.format.plugin.AbstractFileSource;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.Job;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * Performs schema detection on a file path with a given input format.
+ */
+public class SchemaDetector {
+  private final String formatName;
+  private final ValidatingInputFormat inputFormat;
+
+  public SchemaDetector(String formatName, ValidatingInputFormat inputFormat) {
+    this.formatName = formatName;
+    this.inputFormat = inputFormat;
+  }
+
+  /**
+   * Detect the schema for files at a specified path
+   *
+   * @param path path to the files. Can be a file or a directory
+   * @param formatContext format context
+   * @param fileSystemProperties any properties that need to be set in order to read from the Hadoop FileSystem
+   * @return the schema of files at the path, or null if the schema is not known.
+   * @throws IOException
+   */
+  @Nullable
+  public Schema detectSchema(String path, FormatContext formatContext,
+                             Map<String, String> fileSystemProperties) throws IOException {
+    if (path == null) {
+      return null;
+    }
+
+    Job job = JobUtils.createInstance();
+    Configuration configuration = job.getConfiguration();
+    for (Map.Entry<String, String> entry : fileSystemProperties.entrySet()) {
+      configuration.set(entry.getKey(), entry.getValue());
+    }
+    for (Map.Entry<String, String> entry : inputFormat.getInputFormatConfiguration().entrySet()) {
+      configuration.set(entry.getKey(), entry.getValue());
+    }
+
+    // hack: the file path should be passed down to the format, which can decide whether to skip it or not
+    String regexPathFilter = null;
+    if ("avro".equalsIgnoreCase(formatName)) {
+      regexPathFilter = ".+\\.avro$";
+    } else if ("parquet".equalsIgnoreCase(formatName)) {
+      regexPathFilter = ".+\\.parquet$";
+    }
+
+    Path fsPath = new Path(path);
+    // need this to load the extra class loader to avoid ClassNotFoundException for the file system
+    FileSystem fs = JobUtils.applyWithExtraClassLoader(job, getClass().getClassLoader(),
```

Comment:
is this still needed if this happens on the source?

Reply:
This is still required actually, the conf's classloader still doesn't have access to something like GoogleHadoopFileSystem